### PR TITLE
feat(drv): Implement Arduino SD driver

### PR DESCRIPTION
**Pull request by PierreRambaud**
### Description of the feature or fix

Implement Arduino SD driver based on https://www.arduino.cc/reference/en/libraries/sd/ which is compatible with all architectures so you should be able to use it on all the Arduino boards.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
Looks good to me so far! Please, don't forget to update the docs too.

@zjanosy could you try it out?

**Comment by PierreRambaud** [comment]
Not completely working as expected on my side, I need to test more things before putting it on review. And don't worry, I'll update the doc for sure ^^

**Comment by kisvegabor** [comment]
Alright, thank you! :slightly_smiling_face:

**Comment by PierreRambaud** [comment]
@kisvegabor it's ready :)

**Comment by kisvegabor** [comment]
Awesome! :confetti_ball: 

@zjanosy It's ready! :slightly_smiling_face: Please test it on an Arduino board with SD card. Just copy a PNG file into a folder on the SD card and try to display it with an `lv_image` widget.

**Comment by zjanosy** [comment]
I am testing it...

**Comment by zjanosy** [comment]
@PierreRambaud I'm still fighting with the Arduino environment. Could you please add a complete example how to use it? E.g., load and display a .png file. Thanks!

**Comment by PierreRambaud** [comment]
> @PierreRambaud I'm still fighting with the Arduino environment. Could you please add a complete example how to use it? E.g., load and display a .png file. Thanks!

Gonna give you a whole example this week. Sorry for the delay, I've had some health issues recently 😅

**Comment by PierreRambaud** [comment]
Here's a [repository](https://github.com/PierreRambaud/lvgl-arduino-sd-example) compatible with M5Core2, based on platformio and targets this branch.

Don't hesitate if there is something I can do for you

Here's the result I got.

**Comment by kisvegabor** [comment]
It seems [arduino-sd.cpp](https://github.com/PierreRambaud/lvgl-arduino-sd-example/blob/master/src/arduino-sd.cpp) is modified. Is it possible to use an Arduino SD card library as it is?

**Comment by PierreRambaud** [comment]
SD.h is used by default and integrated inside Arduino.h if I correctly understand what you need

File: src/libs/fsdrv/lv_fs_arduino_sd.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+ * @param drv       pointer to a driver where this function belongs
+ * @param file_p    pointer to a file_t variable. (opened with fs_open)
+ * @return          LV_FS_RES_OK: no error or  any error from @lv_fs_res_t enum
+ */
+static lv_fs_res_t fs_close(lv_fs_drv_t * drv, void * file_p)
+{
+    LV_UNUSED(drv);
+    SdFile * lf = (SdFile *)file_p;
+    lf->file.close();
+    lv_free(lf);
```
Comment (FASTSHIFT) [inline src/libs/fsdrv/lv_fs_arduino_sd.cpp]: Why not `delete lf`’?

Comment (PierreRambaud) [reply src/libs/fsdrv/lv_fs_arduino_sd.cpp]: As far as I know, regarding the `lv_free` code, it depends on what's used behind it. It calls `lv_free_core` which can be used by `clib`, `micropython`, or `rtthread`.
It has been asked on another PR to prefer to use `lv_free` over `delete` I think for this reason.

Comment (kisvegabor) [reply src/libs/fsdrv/lv_fs_arduino_sd.cpp]: As the file was allocated with `new` [here](https://github.com/lvgl/lvgl/pull/5968/files#diff-ff738ff520e493610c7223eaebc4ca92c9802377c56e89d0187ce0d1433731caR87) I agree that `delete` shall be used here.

Comment (PierreRambaud) [reply src/libs/fsdrv/lv_fs_arduino_sd.cpp]: Did it in my last squash, really sorry for missing that :)

**Review by FASTSHIFT — APPROVED** [review]

**Comment by kisvegabor** [comment]
We have tried to test it internally but it turned out that we don't have a board with an SD card on which we could compile LVGL + SD under Arduino or PlatfomIO.

Please confirm that ideally all the user needs to do is enabling the driver in `lv_conf.h` and setting `LV_FS_ARDUINO_SD_CS_PIN`.

**Comment by PierreRambaud** [comment]
> We have tried to test it internally but it turned out that we don't have a board with an SD card on which we could compile LVGL + SD under Arduino or PlatfomIO.
> 
> Please confirm that ideally all the user needs to do is enabling the driver in `lv_conf.h` and setting `LV_FS_ARDUINO_SD_CS_PIN`.

This is exactly what I've done with my repository example (expect the M5Core2 implementation to be able to display the png stored inside the SD card).
I confirm the aim behind this PR is to be able to use an SD card based on the Arduino `SD.h` implementation, by only configuring the `lv_conf.h` with `LV_FS_ARDUINO_SD_CS_PIN` and `LV_FS_ARDUINO_SD_FREQUENCY`.

**Review by kisvegabor — APPROVED** [review]

**Comment by kisvegabor** [comment]
I see, thank you! I'm sure you tested it and it works too. I believe others will report the issues if any :slightly_smiling_face:

**Comment by PierreRambaud** [comment]
And I'll be glad to help if I can do anything. Thank you 🙏

**Pull request merged**